packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts

```diff
@@ -0,0 +1,228 @@
+            endpointName: (props.endpointName) ? props.endpointName : undefined,
+        });
+        this.endpointName = endpoint.attrEndpointName;
+    }
+
+    /**
+     * Add production variant to the endpoint configution.
+     *
+     * @param productionVariant: The production variant to add.
+     */
+    public addProductionVariant(productionVariant: ProductionVariant): void {
+        this.productionVariants.push(productionVariant);
+    }
+
+    protected validate(): string[] {
+        const result = super.validate();
+        // check we have at least one production variant
+        if (this.productionVariants.length === 0) {
+            result.push("Must have at least one Production Variant");
+        }
+
+        // check instance count is greater than zero
+        this.productionVariants.forEach(v => {
+            if (v.initialInstanceCount && v.initialInstanceCount < 1) {
+                result.push("Must have at least one instance");
+            }
+        });
+
+        // check variant weight is not negative
+        this.productionVariants.forEach(v => {
+            if (v.initialVariantWeight && v.initialVariantWeight < 0) {
+                result.push("Cannot have negative variant weight");
+            }
+        });
+
+        // validate the instance type
+        this.productionVariants.forEach(v => {
+            if (v.instanceType) {
+                this.validateInstanceType(v.instanceType.toString(), result);
+            }
+        });
+        return result;
+    }
+
+    /**
+     * Render the list of production variants. Set default values if not defined.
+     * @param productionVariants the variants to render
+     */
+    private renderProductionVariants(productionVariants: ProductionVariant[]): CfnEndpointConfig.ProductionVariantProperty[] {
+        return productionVariants.map(v => (
+            {
+                // get the initial instance count. Set to 1 if not defined
+                initialInstanceCount: (v.initialInstanceCount) ? v.initialInstanceCount : 1,
+                // get the initial variant weight. Set to 100 if not defined
+                initialVariantWeight: (v.initialVariantWeight) ? v.initialVariantWeight : 100,
+                // get the instance type. Set to 'ml.c4.xlarge' if not defined
+                instanceType: 'ml.' + ((v.instanceType) ? (v.instanceType.toString()) :
+                    ec2.InstanceType.of(ec2.InstanceClass.C4, ec2.InstanceSize.XLARGE)),
+                modelName: v.model.modelName,
+                // get the variant name. Set to the model name if not defined
+                variantName: (v.variantName) ? v.variantName : v.model.modelName,
+            }
+        ));
+    }
+
+    /**
+     * Validates the provided instance type.
+     * @param instanceType the instance type of the SageMaker endpoint production variant.
+     */
+    private validateInstanceType(instanceType: string, result: string[]) {
```

Comment:
Just have this return `string[]`, and merge the errors in the client of this method, instead of mutating a parameter. I think keeping methods pure is simply easier.